crates/ty_python_semantic/src/semantic_index/builder.rs

```diff
@@ -295,6 +295,15 @@ impl<'db, 'ast> SemanticIndexBuilder<'db, 'ast> {
         // If the scope that we just popped off is an eager scope, we need to "lock" our view of
         // which bindings reach each of the uses in the scope. Loop through each enclosing scope,
         // looking for any that bind each place.
+        // TODO: Bindings in eager nested scopes also need to be recorded. For example:
+        // ```python
+        // class C:
+        //     x: int | None = None
+        // c = C()
+        // class _:
+        //     c.x = 1
+        // reveal_type(c.x)  # revealed: Literal[1]
+        // ```
```

Comment:
I'd like to understand this limitation better. Can we add this as a MD test instead to document what is missing here? If I understand correctly, it would not lead to incorrect types, it's just a limitation in narrowing?

Reply:
Here's a test: https://github.com/mtshiba/ruff/blob/attr-subscript-narrowing/crates/ty_python_semantic/resources/mdtest/narrow/complex_target.md#L39-L45

This is a kind of narrowing specific to attribute/subscript (names defined in a class do not affect the outer scopes). I think it's possible to support this, and in fact pyright can narrow the type correctly. But it requires changes to the eager snapshot mechanism, so I would like to make this a future work.